### PR TITLE
Data refactor

### LINKED ISSUE #18: Distinguish NAIP resolution
Depending on the output resolution and model architecture, we may need to adjust the input NAIP resolution. Currently, the notebooks and naming formats treat the footprint size in meters and in NAIP pixels interchangeably with the assumption that NAIP will always be 1m. To allow for adjusting NAIP resolution, we should distinguish those two parameters throughout. Additionally, we should add `scale` parameters to the `Acquisition.load_*` methods to simplify adjusting NAIP and LiDAR resolution.

### PR DESCRIPTION
**Pull request by aazuspan**
Hopefully the last major refactor! The primary goal here is to fully define each of the generated file types by name to avoid any confusion as we start adjusting more parameters (e.g. NAIP and LiDAR resolution), and to set up a more foolproof system that connects the generation and loading of files to remove the risk of manually entering incorrect parameters (very easy with the previous system).

There are now 4 core classes that are responsible for defining, generating, and loading a corresponding data file:

- `Acquisition`: A pair of LiDAR and NAIP images in Earth Engine. Used to load server-side mosaics.
- `TrainingDataset`: An `Acquisition` that has been sampled into footprints for training. Used to load and sample the acquisition while generating the output CSV, and to load training, testing, and validation sets from the resulting HDF.
- `ModelRun`: A `keras` model with the corresponding `TrainingDataset` that was used to train it, as well as parameters like bands and labels. Used to define, export, and load models.
- `NAIPTFRecord`: A NAIP mosaic stored as a TFRecord. Used both for exporting an `ee.Image` to TFRecord and for loading and predicting on the resulting GeoTIFF.

@grovduck Don't feel like you need to re-run everything or dive too deep into the details (I'm considering this a prototype system that will probably get refined as I start working more with the higher resolution data and get an idea of what's working and what isn't), but if you have any thoughts on the overall design, I'd be interested to hear them! I worry I might have gone a little overboard and made some monster classes with too much responsibility, but I wanted to leave as little room for human error as possible in the usage (for my own sake). I'm also not sure about the choice to try serializing every important attribute into the filename, but the only other alternative I could think of was to set up some kind of database system, which felt like overkill. 

Closes #18

### CODE REVIEW

**Review by grovduck — COMMENTED** [review]
Hey @aazuspan, another great PR.  I ended up going through all four notebooks again and the flow worked great.  As you said, a couple of the classes (particularly `TrainingDataset` and `NAIPTFRecord`) are pretty loaded with functionality.  My only thought is that if you see this package serving more needs than just our immediate ones in the lab, it might be worth strategizing about how those might be broken up.  But, for now, and with our goals in mind, it seems like what you have here allows for the multi-resolution data flows that you were looking for.

Please let me know if any of my comments don't make sense.

File: src/naip_cnn/data.py [hunk 1]
```diff
@@ -116,46 +130,121 @@ def _load(self, shuffle: bool = True, seed: int = 0) -> tf.data.Dataset:
...
+        spacing_str = float_to_str(self.spacing)
+        return "-".join(
+            [
+                self.acquisition.name,
+                naip_res_str,
+                lidar_res_str,
+                footprint_str,
+                spacing_str,
+            ]
+        )
```
Comment (grovduck) [inline src/naip_cnn/data.py:191]: Yep, as you say, kind of a beastly file naming convention, but I think all attributes are needed to differentiate between possible parameters.  The only other thought I had was to have a sidecar file (JSON or YAML) that would store some of these attributes, but that becomes difficult when you're using this as a download name to kick off a GEE export and presumably you'd have to write this sidecar file locally, so they would become disconnected.  I don't have too much to offer on this one.

File: notebooks/03_export_naip.ipynb [hunk 2]
```diff
@@ -45,64 +44,33 @@
...
     "            [-119.37239004502233, 44.81785572318615],\n",
     "        ]\n",
-    "    ],\n",
-    "    None,\n",
-    "    False,\n",
-    ")"
+    "    ]\n",
+    ")\n",
+    "\n",
+    "naip = NAIPTFRecord(prefix=\"MAL\", footprint=(150, 150), res=1, year=2016, bounds=BBOX)"
```
[the remark this answers is not in the document]
Comment (aazuspan) [reply notebooks/03_export_naip.ipynb]: Good catch! Last minute naming change that I forgot to apply here.

File: src/naip_cnn/acquisitions.py [hunk 3]
```diff
@@ -35,50 +43,99 @@ def mask(self):
             .reproject(self.proj.atScale(30))
```
[the remark this answers is not in the document]
Comment (aazuspan) [reply src/naip_cnn/acquisitions.py:43]: Yeah, sticking with the LCMS scale was the idea. I was pretty sure this would work as you're describing, but I'm glad it sounds right to you as well!

File: src/naip_cnn/config.py [hunk 4]
```diff
@@ -1,3 +1,14 @@
+from pathlib import Path
+
+# The Earth Engine path where assets are stored
+EE_ASSET_DIR = "projects/ee-maptheforests/assets"
```
[the remark this answers is not in the document]
Comment (aazuspan) [reply src/naip_cnn/config.py:4]: Mostly just something to tidy up! The band names were lost when I ingested the images manually (if you know of any workaround to avoid that, let me know!), so I used a Code Editor script to re-set the band names and re-ingest them into a collection. It's a bit of a pain, but I'm assuming we're not going to be adding new training data frequently enough to worry about a better system.

File: src/naip_cnn/data.py (hunk 1 above)
Comment (aazuspan) [reply src/naip_cnn/data.py:191]: A sidecar JSON file could definitely work, but as you mentioned there's some complexity with how you name the exported data. I'm hoping this is as long as the file names will have to get for the training datasets, although for the `ModelRun` in particular I'm a little worried about how we'll handle runs with different hyperparameters. I'll keep the sidecar idea in mind and may have to resort to that if things get too unwieldy in the future.

File: notebooks/02_training.ipynb
[the remark this answers is not in the document]
Comment (aazuspan) [reply notebooks/02_training.ipynb:1]: Fair point! I've noticed the same thing you describe where the LiDAR really doesn't seem to match the imagery while the prediction looks pretty good, so I think this is a good distinction to make.

I didn't re-run the notebook to regenerate the figure, but I did fix the code so that it will be correct next run.

File: src/naip_cnn/config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (grovduck) [reply src/naip_cnn/config.py:4]: Makes sense to me!  (Sorry, I don't have anything off the top of my head to make band names stick.  I'm always impressed when *exported* TIFs from GEE maintain those band names, so surely there is a way to do it with GDAL, but I haven't researched this fully.)

**Comment by aazuspan** [comment]
Thanks for giving this a look @grovduck, and going to the trouble of re-running everything! I agree with your concerns about `TrainingDataset` and `NAIPTFRecord`, and I'm also starting to think that serializing everything into filenames isn't going to be very scalable, especially when it comes to complex model hyperparameters. I think a database solution is probably the "correct" way to handle a lot of this, but I definitely don't want to tackle building that from scratch, so I'm on the lookout for premade solutions that might accomplish what we need. [MLflow](https://mlflow.org/docs/latest/introduction/index.html) is looking particularly interesting, but I need to spend a little more time figuring out its capabilities.

I'll pause this for now while I look into if/how that could be integrated, and keep you posted!

**Comment by grovduck** [comment]
> I'll pause this for now while I look into if/how that could be integrated, and keep you posted!

Sounds good, @aazuspan.  Both of those packages look interesting, but probably also a time commitment to get up to speed!

**Comment by aazuspan** [comment]
@grovduck Finally getting back around to this! For now, I decided that the simplest path forward is to clean up this PR and get it merged, then try and integrate DVC from that point. I have a rough plan for how DVC can be integrated after experimenting with a few toy projects, which I'll outline in a new issue. Bottom line, I think the best approach will be to patch in DVC at the training stage, using it to record each combination of dataset and model parameter with the associated model metrics and outputs. That will undo a few of the changes here, but trying to shoehorn it into this PR seemed like it might stall this even longer.

For now, the biggest change was to split up the old `TrainingDataset` class into a `_NAIPHDF5Dataset` that handles the NAIP-specific tensor processing/loading and a `NAIPDatasetWrapper` that wraps that class, the `Acquisition`, and all relevant metadata. I think you described the issue with the class very well here:

> it feels like TrainingDataset should have a HDF5Dataset rather than be one

so that's what I attempted to implement with the new wrapper class. I made a few other changes to try and clean up the class hierarchy and turned the HDF5-specific loading functionality into a mixin, with the idea that it could be swapped out if there was another data format that loaded in a different way. 

I reran the sampling and training notebooks, and everything seems to be working as expected! I also keep accidentally commiting `autoreload` blocks in my notebooks, so I added a script that enables that by default by setting the Docker-level user config (it shouldn't affect your local config since that's not mounted in the Docker image).

File: src/naip_cnn/data.py [hunk 5]
```diff
@@ -116,46 +130,121 @@ def _load(self, shuffle: bool = True, seed: int = 0) -> tf.data.Dataset:
...
 
 
-class NAIPDataset(HDF5Dataset):
-    """A dataset of NAIP images and associated forest attribute labels."""
+class TrainingDataset(HDF5Dataset):
+    """A dataset of NAIP images and associated forest attribute labels.
+
+    This class is used both to define and generate sample datasets and to load
+    previously generated datasets.
+    """
```
[the remark this answers is not in the document]
Comment (aazuspan) [reply src/naip_cnn/data.py]: Great feedback, thank you! I tried to address the first issue in my last comment.

> Then it seems like we'd want a superclass of Acquisition that would know how to get paired sets of imagery - one used for covariates and one used for labels.

This is a very good point - the idea of a paired set of predictor and label images certainly isn't unique to this application, and could come in handy down the line. I think I'm going to punt on this for now to keep things moving with our short-term use case, but will plan to revisit this when needed.

[the remark this answers is not in the document]
Comment (grovduck) [reply src/naip_cnn/data.py]: Totally fair to push this one off right now.

**Comment by aazuspan** [comment]
> I'm hoping DVC might be useful in other applications as well for model metadata.

Definitely! I've only scratched the surface following a few tutorials, but it looks like a pretty powerful tool. 

> would you create an abstract base class for _NAIPHDF5Dataset (call it _NAIPDataset) and then have other derived classes like (stupidly) _NAIPGeoTiffDataset implement the required methods? Then the dataset attribute in NAIPDatasetWrapper would actually be any derived class that subclasses _NAIPDataset?

I hadn't fully thought this through, but what I had in mind was having other mixins to handle loading (e.g. `_GeoTiffDatasetMixin`). A `_NAIPGeoTiffDataset` could inherit from that, but so could a `_LandsatGeoTiffDataset`, for example. I think what you're describing with a shared `_NAIPDataset` makes perfect sense though, and would be a good way to share NAIP-specific functionality independent of the loading mechanism.

> This goes to show you that I didn't even realize that this was an issue 😄!

I think the only real issue is having to make embarrassing "remove code I accidentally left behind" commits, but having it autoreload will hopefully also save me some headscratching when I can't figure out why my changes aren't having any effect on the notebooks!

Thanks for your help and patience on this one! I know reviewing those notebook diffs with 180k changed lines aren't much fun, but I definitely needed someone to call out my questionable design choices from the first pass!
